src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/validation/HaFlowValidationFsm.java

```diff
@@ -0,0 +1,340 @@
+        super(NEXT, ERROR, commandContext, carrier, eventListeners);
+        this.flowId = flowId;
+
+        service = new HaFlowValidationService(persistenceManager, new RuleManagerImpl(ruleManagerConfig));
+    }
+
+    protected void receiveData(State from, State to,
+                               Event event, Object context) {
+        HaFlow haFlow;
+        try {
+            haFlow = service.findValidHaFlowById(flowId);
+        } catch (FlowNotFoundException e) {
+            log.error("The haFlow {} has not been found when sending commands to the SpeakerWorkerBolt", flowId, e);
+            sendException(e.getMessage(), "Receiving rules operation in HaFlowValidationFsm", ErrorType.NOT_FOUND);
+            return;
+        } catch (IllegalFlowStateException e) {
+            log.error("Could not validate the haFlow: the haFlow {} is in DOWN state", flowId, e);
+            sendException("Could not validate the haflow",
+                    format("Could not validate the haFlow: haFlow %s is in DOWN state", flowId),
+                    ErrorType.UNPROCESSABLE_REQUEST);
+            return;
+        }
+
+        switchIdSet.addAll(extractSwitchIdSet(haFlow));
+
+        awaitingRulesSize = switchIdSet.size();
+        log.debug("Send commands to get the rules on the switches");
+        switchIdSet.forEach(switchId ->
+                getCarrier().sendSpeakerRequest(flowId, new DumpRulesForFlowHsRequest(switchId)));
+
+        Set<SwitchId> meterSwitchIds = new HashSet<>(haFlow.getEndpointSwitchIds());
+        meterSwitchIds.add(haFlow.getForwardPath().getYPointSwitchId());
+        Optional.ofNullable(haFlow.getProtectedForwardPath()).ifPresent(e -> meterSwitchIds.add(e.getYPointSwitchId()));
+
+        log.debug("Send commands to get the meters on the termination switches");
+        awaitingMetersSize = meterSwitchIds.size();
+        meterSwitchIds.forEach(switchId -> {
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(switchId));
+        });
+
+        log.debug("Send commands to get the groups on the termination switches");
+        awaitingGroupsSize = 2;
+
+        getCarrier().sendSpeakerRequest(flowId,
+                new DumpGroupsForFlowHsRequest(haFlow.getForwardPath().getYPointSwitchId()));
+        Optional.ofNullable(haFlow.getProtectedForwardPath()).ifPresent(protectedFlowPath -> {
+            getCarrier().sendSpeakerRequest(flowId,
+                    new DumpGroupsForFlowHsRequest(protectedFlowPath.getYPointSwitchId()));
+        });
+
+
+    }
+
+    protected void receivedRules(State from, State to,
+                                 Event event, Object context) {
+        FlowDumpResponse flowDumpResponse = (FlowDumpResponse) context;
+        log.info("The Switch rules received for the switch {}",
+                flowDumpResponse.getFlowSpeakerData().stream().findFirst().map(e -> e.getSwitchId().toString())
+                        .orElse("No switch Id"));
```

Comment:
You have switch id in FlowDumpResponse. You can use it.

Reply:
fixed